File: src/main/java/com/learnspark/submission/service/SubmissionService.java
package com.learnspark.submission.service;

import com.learnspark.common.exception.BusinessException;
import com.learnspark.common.exception.ErrorCode;
import com.learnspark.gamification.service.ScoreService;
import com.learnspark.knowledge.entity.KnowledgeEntry;
import com.learnspark.knowledge.repository.KnowledgeEntryRepository;
import com.learnspark.plan.entity.Task;
import com.learnspark.plan.repository.TaskRepository;
import com.learnspark.plan.service.TaskService;
import com.learnspark.submission.dto.ReviewResult;
import com.learnspark.submission.dto.SubmissionRequest;
import com.learnspark.submission.dto.SubmissionResponse;
import com.learnspark.submission.entity.Submission;
import com.learnspark.submission.repository.SubmissionRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import java.time.LocalDateTime;
import java.util.List;

/**
 * 提交审核服务。
 *
 * <p>提交流程：校验任务归属 → 创建提交记录 → AI 审核（DeepSeek 真实 / Mock 降级）→ 更新任务状态 → 写入知识库。
 * 审核策略：优先调用 DeepSeek 真实审核；未配置 Key 或调用失败时降级到 Mock（passed=true, score=9）。
 * Key 无效时不降级，抛 AI_KEY_INVALID 提示用户修正配置。
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class SubmissionService {

    private final SubmissionRepository submissionRepository;
    private final KnowledgeEntryRepository knowledgeEntryRepository;
    private final TaskRepository taskRepository;
    private final TaskService taskService;
    private final DeepSeekReviewService deepSeekReviewService;
    private final ScoreService scoreService;

    /**
     * 提交任务并触发 AI 审核。
     *
     * @param userId  当前用户 ID
     * @param taskId  任务 ID
     * @param request 提交内容
     * @return 提交响应（含审核结果与知识条目 ID）
     */
    @Transactional
    public SubmissionResponse submit(String userId, String taskId, SubmissionRequest request) {
        // 1. 校验任务归属
        Task task = taskService.getTaskForUser(userId, taskId);

        // 2. 已通过的任务不允许重复提交
        if ("passed".equalsIgnoreCase(task.getStatus())) {
            throw new BusinessException(ErrorCode.TASK_ALREADY_PASSED);
        }

        // 3. AI 审核：优先使用 DeepSeek 真实审核，未配置 Key 或调用失败时降级到 Mock
        ReviewResult review = doReview(userId, task, request.getContent());

        // 4. 创建提交记录
        Submission submission = Submission.builder()
                .taskId(taskId)
                .userId(userId)
                .content(request.getContent())
                .attachmentUrls(request.getAttachmentUrls())
                .aiFeedback(review.feedback())
                .aiScore(review.score())
                .passed(review.passed())
                .aiModel(review.model())
                .aiRawResponse(review.rawResponse())
                .reviewedAt(LocalDateTime.now())
                .build();
        submission = submissionRepository.save(submission);
        log.info("任务提交成功: taskId={}, submissionId={}, passed={}", taskId, submission.getId(), review.passed());

        // 5. 更新任务状态
        if (review.passed()) {
            task.setStatus("passed");
            task.setCompletedAt(LocalDateTime.now());
        } else {
            task.setStatus("submitted");
        }
        taskRepository.save(task);

        // 5.5 通过则加分并更新连续打卡（触发徽章检查）
        if (review.passed()) {
            scoreService.awardOnPass(userId, task.getProjectId(), review.score());
        }

        // 6. 将提交内容写入知识库
        KnowledgeEntry entry = createKnowledgeEntryFromSubmission(userId, task, submission);
        knowledgeEntryRepository.save(entry);
        log.info("知识条目已创建: entryId={}, sourceType=submission, sourceId={}", entry.getId(), submission.getId());

        return SubmissionResponse.from(submission, task.getStatus(), entry.getId());
    }

    /**
     * 查询任务的历史提交记录。
     */
    @Transactional(readOnly = true)
    public List<Submission> listByTask(String userId, String taskId) {
        // 校验任务归属
        taskService.getTaskForUser(userId, taskId);
        return submissionRepository.findByTaskIdOrderBySubmittedAtDesc(taskId);
    }

    /**
     * 执行 AI 审核：优先 DeepSeek 真实审核，降级到 Mock。
     *
     * <p>降级场景：
     * <ul>
     *   <li>未配置 API Key / 本地模式 → Mock</li>
     *   <li>调用超时 / 网络错误 / 限流 → Mock</li>
     *   <li>Key 无效 → 抛 AI_KEY_INVALID（不降级，需用户修正配置）</li>
     * </ul>
     */
    private ReviewResult doReview(String userId, Task task, String content) {
        ReviewResult result = deepSeekReviewService.review(userId, task, content);
        if (result != null) {
            return result;
        }
        return mockReview();
    }

    /**
     * Mock 审核：固定返回通过、9 分。
     *
     * <p>当用户未配置 DeepSeek Key 或调用失败时作为降级方案。
     */
    private ReviewResult mockReview() {
        return new ReviewResult(
                true,
                9,
                "（Mock 模式）自动审核通过，请继续加油！",
                "mock",
                null
        );
    }

    /** 从提交内容创建知识条目 */
    private KnowledgeEntry createKnowledgeEntryFromSubmission(String userId, Task task, Submission submission) {
        String title = StringUtils.hasText(task.getTitle())
                ? "任务提交：" + task.getTitle()
                : "任务提交：" + task.getDescription().substring(0, Math.min(30, task.getDescription().length()));

        return KnowledgeEntry.builder()
                .userId(userId)
                .projectId(task.getProjectId())
                .title(title)
                .content(submission.getContent())
                .contentMd(submission.getContent())
                .sourceType("submission")
                .sourceId(submission.getId())
                .wordCount(countWords(submission.getContent()))
                .parseStatus("done")
                .build();
    }

    /** 统计字数（按非空白字符计数，适配中英文混合） */
    private int countWords(String content) {
        if (content == null || content.isBlank()) {
            return 0;
        }
        return content.replaceAll("\\s+", "").length();
    }
}

File: learnspark-kmp/server/src/main/kotlin/com/learnspark/server/service/SubmissionService.kt
package com.learnspark.server.service

import com.learnspark.server.domain.entity.Submission
import com.learnspark.server.domain.repository.SubmissionRepository
import com.learnspark.server.domain.repository.TaskRepository
import com.learnspark.server.service.ai.AiConfigService
import com.learnspark.server.service.ai.OpenAICompatibleClient
import com.learnspark.server.service.ai.SubmissionReviewer
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.UUID

/**
 * R2：任务提交服务（含 AI 评审）。
 *
 * 写流程：
 * 1) create(content) → 立即返回 submission（不阻塞）
 * 2) 异步 @Async 触发 AI 评审
 * 3) DeepSeek 返回 → 写回 ai_score/ai_feedback/ai_highlights/reviewed_at
 *
 * 读流程：
 * - listByTask / get
 * - 跨级 ownership：submission → task → phase → project → user
 */
@Service
class SubmissionService(
    private val submissionRepository: SubmissionRepository,
    private val taskRepository: TaskRepository,
    private val phaseService: PhaseService,
    private val aiConfigService: AiConfigService,
    private val submissionReviewer: SubmissionReviewer,
) {
    private val log = LoggerFactory.getLogger(SubmissionService::class.java)

    fun listByTask(taskId: String, userId: String): List<Submission> {
        if (!taskBelongsToUser(taskId, userId)) return emptyList()
        return submissionRepository.findByTaskId(taskId)
            .sortedByDescending { it.updatedAt }
    }

    fun get(id: String, userId: String): Submission? {
        val s = submissionRepository.findById(id).orElse(null) ?: return null
        return if (taskBelongsToUser(s.taskId, userId)) s else null
    }

    @Transactional
    fun create(
        taskId: String,
        userId: String,
        content: String,
        triggerReview: Boolean = true,
    ): CreateResult {
        if (!taskBelongsToUser(taskId, userId)) return CreateResult.TaskNotFound
        val submission = Submission(
            id = UUID.randomUUID().toString(),
            taskId = taskId,
            userId = userId,
            content = content,
            version = 1L,
        )
        val saved = submissionRepository.save(submission)
        log.info("submission created id={} user={} task={}", saved.id, userId, taskId)
        if (triggerReview) triggerAiReviewAsync(saved.id, userId)
        return CreateResult.Ok(saved)
    }

    @Transactional
    fun update(
        id: String,
        userId: String,
        content: String? = null,
        baseVersion: Long? = null,
    ): UpdateResult {
        val existing = get(id, userId) ?: return UpdateResult.NotFound
        if (baseVersion != null && existing.version != baseVersion) {
            return UpdateResult.VersionConflict(existing.version, existing)
        }
        content?.let { existing.content = it }
        existing.version = existing.version + 1
        val saved = submissionRepository.save(existing)
        return UpdateResult.Ok(saved)
    }

    @Transactional
    fun delete(id: String, userId: String): Boolean {
        val existing = get(id, userId) ?: return false
        submissionRepository.delete(existing)
        return true
    }

    /**
     * 异步触发 AI 评审：
     * - 用户未配置 / 禁用 DeepSeek → 静默跳过（submission.reviewed_at 不更新）
     * - DeepSeek 错误 → reviewed_at = now + ai_feedback = error message
     * - 成功 → 写回 score/feedback/highlights/reviewed_at
     */
    @Async
    fun triggerAiReviewAsync(submissionId: String, userId: String) {
        try {
            submissionReviewer.review(submissionId, userId)
        } catch (e: Exception) {
            log.warn("AI review background failed submission={}: {}", submissionId, e.message)
        }
    }

    private fun taskBelongsToUser(taskId: String, userId: String): Boolean {
        val task = taskRepository.findById(taskId).orElse(null) ?: return false
        return phaseService.get(task.phaseId, userId) != null
    }

    sealed class CreateResult {
        data class Ok(val submission: Submission) : CreateResult()
        data object TaskNotFound : CreateResult()
    }

    sealed class UpdateResult {
        data class Ok(val submission: Submission) : UpdateResult()
        data object NotFound : UpdateResult()
        data class VersionConflict(val currentVersion: Long, val current: Submission) : UpdateResult()
    }
}
